src/control/cmd/dmg/utils.go

```diff
@@ -26,25 +26,124 @@ package main
 import (
 	"fmt"
 	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/pkg/errors"
 
 	"github.com/daos-stack/daos/src/control/client"
+	"github.com/daos-stack/daos/src/control/lib/hostlist"
 )
 
-func hasConns(results client.ResultMap) (bool, string) {
-	out := sprintConns(results)
-	for _, res := range results {
-		if res.Err == nil {
-			return true, out
+type HostGroup struct {
```

Comment:
I'm curious about why you moved HostGroup away from the map[string]*hostlist.HostSet implementation. It seemed like a generally-useful structure to coalesce hosts into groups with some string-based commonality. Now it's only useful for this one narrow use case in this part of dmg. Just seems like a missed opportunity.

Reply:
Passing around a map meant that I had to worry about sorting keys each time I wanted to create a slice. the code ends up simpler if I keep the map usage within the function and returns sorted slices (less "slicing and dicing" as you put it). I also then don't have to worry about order of iteration through a map either in tests or code.

Reply:
OK. As an exercise, though, what would it look like if you were to extend the work in this PR to handle coalescing hosts by format result in addition to connection status? In other words, if you had two groups of hosts, one where the format succeeded and one where it didn't, how would you use the code you've written to display that?

Reply:
It's been a localised trade-off, the re-factor makes this code simpler and neater. I didn't want to prematurely optimise for a generic use case when we don't have a need for one. What you are proposing in your "exercise" is a significant design change where we move from a model establishing connection state before iterating over those connections to perform actions, to, establishing connection and performing action in the same iteration of the loop. That change in design would obviously require a bit of a refactor and designing from the beginning with that kind of future change in mind is exactly what people are wary of when they talk about "premature optimisation". In this situation, if you wanted to filter by format result in addition to connection status then I would create a type similar to HostGroupsError e.g. HostGroupsFormat which would contain fields HostGroups and state. Simple filter/sort functions could then be used to display. I'm not sure would be in favour of that design change as decoupling connection establishment from performing operations on the connected servers seems like an important and valuable separation of concerns.

Reply:
My proposal wasn't meant to be taken wholesale, I just offered some working code to show what I was talking about. Some of the details may not be right for this specific problem, but I don't think it's premature to be thinking about how we can start building tools to solve the problem of handling scalable user interfaces. My concern with your approach is that it seemed like there was a lot of code that would only be used for this specific use case, when with a bit of extra up-front design work we could have something that is used for this solution but could also be used for work that is coming right up afterward. Before we release 1.0, we still need to deal with coalescing output by host groups, e.g.

Format results:
  node-[1-8,10-16]: success
                  node-9: failed

Storage Scan:
                       |   DCPM  |  NVMe  |
  node-[1-16]:    512GiB     2.5TiB

Using a map[string]*HostSet data structure gives us quite a bit of flexibility in how we define the key. A string can hold all kinds of structured data, e.g. csv, json, etc. The simple stringer interface is just a starting point. More advanced formatting can be done as necessary by the code that uses the data structure.

Reply:
I will work your suggestions into the PR